filebeat/input/v2/plugin.go

```diff
@@ -0,0 +1,112 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package v2
+
+import (
+	"github.com/elastic/beats/heartbeat/scheduler"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/go-concert/chorus"
+)
+
+type Plugin struct {
+	Name        string
+	Constraints Constraint
+	Doc         string
+
+	// TODO: config schema info for config validation
+
+	// Configure configures an Input if possible. Returns an error if the
+	// configuration is false.
+	// The input return must not attempt to run or create any kind of connection yet.
+	// The input is only supposed to be a types place holder for the untyped
+	// configuration.
+	Configure func(log *logp.Logger, config *common.Config) (Input, error)
+}
+
+// Input defines the functionality all input plugins must provide.
+type Input interface {
+	InputTester
+	RunnerFactory
+}
+
+// InputTester provides functionality for testing aa feature its configuration without
+// having to run it for a long time.
+// The test is supposed to not take too much time. Packages running a tester
+// might run the test with a pre-configured timeout.
+type InputTester interface {
+	TestInput(closer *chorus.Closer, log *logp.Logger) error
+}
+
+// RunnerFactory creates a runner that can be started and stopped.
+// The context passed to the runner is used to signal beats shutdown.
+//
+// The provided observer must be used to provide report the inputs lifecycle
+// state change.
+type RunnerFactory interface {
+	CreateRunner(ctx Context) Runner
+}
+
+// Runner must run a feature until the run is either stopped, or fails hard
+// making it impossible to recover.
+// Each call to Run must be fully isolated and independent of other runs. No
+// global state should be shared via the Input or global variables.
+// If the Runner is stopped, it is assumed that all resources are freed.
+//
+// NOTE: the type is defined such that it is compatible with libbeat config
+// reloading and autodiscovery.
+type Runner interface {
+	String() string
+	Start()
+	Stop()
+}
+
+// Context provides access to common resources and shutdown signaling to
+// inputs. The `Closer` provided is compatible to `context.Context` and can be used
+// with common IO libraries to unblock connections on shutdown.
+type Context struct {
+	// StoreAccessor allows inputs to access a resource store. The store can be used
+	// for serializing state, but also for coordination such that only one input
+	// collects data from a resource.
+	StoreAccessor
+
+	// Closer provides support for shutdown signaling.
+	// It is compatible to context.Context, and can be used to cancel IO
+	// operations during shutdown.
+	Closer *chorus.Closer
```

Comment:
How about actually using `context.Context`? It would be more friendly for newcomers

Reply:
They have a compatible interface so you can use it directly in anywhere that `context.Context` is available.
The different that the `chorus.Closer` has is `support ripple close of go routine`. It a library that we have created a few months ago to experiment with patterns https://github.com/elastic/go-concert/blob/master/chorus/closeref.go

Reply:
that sounds interesting, Is there any documentation I can look at? I'm not very familiar with "ripple close"

Reply:
both context.Context and chorus.Closer build a hierarchy/tree for passing a closing signal down the chain, but stop if some intermediary has already been closed.

context.Context (besides some exceptions) builds the hierarchy by creating temporary go-routines. This is not optimal, but these go-routines are very short lived, as context.Context is supposed to be somewhat short lived.

The chorus.Closer on the other hand provides some helpers for calling a custom close function while the shutdown signal is propagated (useful for files, low level network sockets or if io.Reader is expected). It works somewhat similar to context.Context, but is supposed to have an extended lifetime. It is ensured that the type can be used when context.Context is expected, so it can be used with any networking libs that support cancelling. The Closer builds a simple tree instead of starting go-routines for signal propagation (reduces memory usage + reduces number of go-routines when creating a trace of all active go-routines).

Reply:
Would it make sense to put `context.Context` on the interface and then pass a `chorus.Closer`? My concern is more about maintainability and documentation. Newcomers seeing `context.Context` will understand it out of the box, while `chorus.Closer` requires some learning.

Reply:
might consider it in the later PRs. Reasons I don't want to use `context.Context` yet:
- chorus package doesn't yet accept an interface. for some functionality. We need to work a little on the package 
- `context.Value` is inviting devs to add hacks to Beats (a context might be used by autodiscovery in the future)
- actually we just want some means of shutdown signaling. The signaling is compatible to `context.Context` only for convenience. We might consider to make remove this compatibility, but provide a 'casting' function for you to use the signaler with `context.Context`